jit/core/jit_cleaner.py

```diff
@@ -0,0 +1,226 @@
+    end_datetime = datetime.strptime(end, "%Y-%m-%dT%H:%M:%S.%fZ").replace(
+      tzinfo=timezone.utc)
+
+    publish_time = received_message.message.publish_time.astimezone(
+      timezone.utc)
+    time_for_now = datetime.now(timezone.utc)
+    if end_datetime < time_for_now:
+        jit_logger.info(f"access expired: {received_message.message}")
+        modify_policy_remove_member(conf, target_project_id, role, member,
+                                    start, end)
+    elif publish_time + timedelta(days=6) < time_for_now:
+        republish(conf.pubsub_topic_name, received_message.message.data,
+                  constant.MessageOrigin.BINDING.value)
+    else:
+        # rest of the cases are not expired cases, we dont do ack
+        return False
+    return True
+
+
+def run_jit_cleaner(conf: config.JitConfig):
+    """
+    jit_cleaner fetches message from pubsub with max=NUM_MESSAGES by loop
+    until it consume all the message in pubsub
+    only process the origin is jit-approval which is stored in pubsub message attribute
+    based on condition start and expire, decide whether need to clean up
+    or republish.
+    https://cloud.google.com/iam/docs/granting-changing-revoking-access#single-role
+    https://cloud.google.com/resource-manager/reference/rest/v1/projects/getIamPolicy
+    https://cloud.google.com/run/docs/tutorials/gcloud
+    :return:
+    """
+
+    jit_logger.info("cleanup process")
+    subscriber = pubsub_v1.SubscriberClient()
+
+    # Wrap subscriber in a 'with' block to automatically call close() when done.
+    with subscriber:
+
+        while True:
+            # The subscriber pulls a specific number of messages. The actual
+            # number of messages pulled may be smaller than max_messages.
+            response = subscriber.pull(
+              request={"subscription": conf.pubsub_subscription_path,
+                       "max_messages": conf.num_messages},
+              retry=retry.Retry(deadline=300),
+            )
+
+            if len(response.received_messages) == 0:
+                jit_logger.info(
+                  f"Received 0 message. PubSub subscription({conf.pubsub_subscription_path}) is empty."
+                )
+                break
+
+            ack_ids = []
+            for received_message in response.received_messages:
+                try:
+                    is_ack = process_pubsub_msg(conf, received_message)
+                    ack_ids += [received_message.ack_id] if is_ack else []
+                except (google.auth.exceptions.MutualTLSChannelError,
+                        google.auth.exceptions.DefaultCredentialsError) as google_ex:
+                    # if google api exception,
+                    # we should not ack it, and make it reappear.
+                    jit_logger.error("process pubsub google api exception: ",
+                                     google_ex)
+                    continue
+                except Exception as e:
```

Comment:
This is dangerous because we will then delete the data so we can't understand how we fix it